services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy

```diff
@@ -0,0 +1,186 @@
+    }
+
+    boolean findDirectLinks(Switch src, Switch dst, List<IslInfoData> links) {
+        def connectingLinks = links.findAll {link ->
+            (link.path[0].switchId == src.dpId) && (link.path[-1].switchId == dst.dpId)
+        }
+        return connectingLinks.empty
+    }
+
+    def getNonNeighbouringSwitches() {
+        def islInfoData = northbound.getAllLinks()
+        def switches = topologyDefinition.getActiveSwitches()
+        def differentSwitches = [switches, switches].combinations().findAll {src, dst -> src.dpId != dst.dpId}
+
+        return differentSwitches.find {src, dst -> findDirectLinks(src, dst, islInfoData)}
+    }
+
+    def getNeighbouringSwitches() {
+        def switches = topologyDefinition.getActiveSwitches()
+        def islInfoData = northbound.getAllLinks()
+        return [switches, switches].combinations().unique().find { src, dst -> !findDirectLinks(src, dst, islInfoData) }
+    }
+
+    def getSingleSwitch() {
+        def switches = topologyDefinition.getActiveSwitches()
+        return [switches.first(), switches.first()]
+    }
+
+    @Unroll
+    def "Flow validation should fail if #rule switch rule is deleted"() {
+        given: "Flow with a transit switch exists"
+
+        def (Switch srcSwitch, Switch dstSwitch) = getNonNeighbouringSwitches()
+        def flow = flowHelper.randomFlow(srcSwitch, dstSwitch)
+        flow = northbound.addFlow(flow)
+        assert flow?.id
+        assert Wrappers.wait(20) { "up".equalsIgnoreCase(northbound.getFlow(flow.id).status) }
+        def forward = database.getFlow(flow.id)?.left
+        def involvedSwitches = pathHelper.getInvolvedSwitches(flow.id)*.dpId
+
+        and: "Rules are installed on all switches"
+        assert involvedSwitches.every{ sw ->
+            northbound.getSwitchRules(sw).flowEntries.find { it.cookie == forward.cookie }
+        }
+        when: "${rule} switch rule is deleted"
+        def dpId = involvedSwitches[item]
+        def cookie = forward.cookie.toString()
+        northbound.deleteSwitchRules(dpId, cookie)
+
+        def validationResult = northbound.validateFlow(flow.id)
+        then: "Forward and reverse flows should be validated"
+        validationResult.size() == 2
+        and: "Reverse flow validation must be successful"
+        validationResult.any { isFlowValid(it) }
+        and: "Forward flow validation should fail"
+        def fwdFlow = validationResult.find { !isFlowValid(it) }
+        and: "Only one rule should be missing"
+        def rules = findRulesDiscrepancies(fwdFlow)
+        rules.keySet().size() == 1
+        rules[dpId.toString()] == cookie
+        cleanup: "Delete the flow"
+        if (flow){
+            northbound.deleteFlow(flow.id)
+        }
+        where:
+        rule << ["ingress", "transit", "egress"]
+        item << [0, 1, -1]
+    }
+
+    @Unroll
+    def "Both flow and switch validation should fail if flow rule is missing with #flowconfig flow configuration"() {
```

Comment:
First test covers only non-neighboring switches, second one tries all 3 possible flow configurations. It is possible to factor 1st test case in such a way, but I'm afraid it'd get even more messy.

Reply:
Cases of deleting ingress/egress rules on flow with transit switches will be omitted if I delete first test. As I said, it could be merged into a single test cases, but what for?

Reply:
> Cases of deleting ingress/egress rules on flow with transit switches will be omitted if I delete first test
true, that makes sense. So does the 'transit' case is duplicating or what is the difference?

Reply:
So what if add to the where block next values
```
        where:
        flowconfig << ["ingress on flow with transit switch", "egress on flow with transit switch"]
        method << [{ getNonNeighbouringSwitches() }, { getNonNeighbouringSwitches() }]
        item << [0, -1]
```
Will it fully cover the 1st test or what are the other differences?

Reply:
Resolved by factoring the test cases